packages/asset-buyer/src/quote_consumers/dynamic_swap_quote_consumer.ts

```diff
@@ -0,0 +1,117 @@
+
+        this._exchangeConsumer = new ExchangeSwapQuoteConsumer(supportedProvider, options);
+        this._forwarderConsumer = new ForwarderSwapQuoteConsumer(supportedProvider, options);
+    }
+
+    public async getCalldataOrThrowAsync(
+        quote: SwapQuote,
+        opts: Partial<DynamicSwapQuoteGetOutputOpts>,
+    ): Promise<CalldataInfo> {
+        assert.isValidSwapQuote('quote', quote);
+        const consumer = await this._getConsumerForSwapQuoteAsync(quote, opts);
+        return consumer.getCalldataOrThrowAsync(quote, opts);
+    }
+
+    public async getSmartContractParamsOrThrowAsync(
+        quote: SwapQuote,
+        opts: Partial<DynamicSwapQuoteGetOutputOpts>,
+    ): Promise<SmartContractParamsInfo<SmartContractParams>> {
+        assert.isValidSwapQuote('quote', quote);
+        const consumer = await this._getConsumerForSwapQuoteAsync(quote, opts);
+        return consumer.getSmartContractParamsOrThrowAsync(quote, opts);
+    }
+
+    public async executeSwapQuoteOrThrowAsync(
+        quote: SwapQuote,
+        opts: Partial<DynamicSwapQuoteExecutionOpts>,
+    ): Promise<string> {
+        assert.isValidSwapQuote('quote', quote);
+        const consumer = await this._getConsumerForSwapQuoteAsync(quote, opts);
+        return consumer.executeSwapQuoteOrThrowAsync(quote, opts);
+    }
+
+    private async _getConsumerForSwapQuoteAsync(
+        quote: SwapQuote,
+        opts: Partial<DynamicSwapQuoteGetOutputOpts>,
+    ): Promise<ValidSwapQuoteConsumer> {
+        const wethAssetData = assetDataUtils.getEtherTokenAssetData(this._contractWrappers);
+        if (swapQuoteConsumerUtils.isValidForwarderSwapQuote(quote, wethAssetData)) {
+            if (opts.takerAddress !== undefined) {
+                assert.isETHAddressHex('takerAddress', opts.takerAddress);
+            }
+            const ethAmount = opts.ethAmount || quote.worstCaseQuoteInfo.totalTakerTokenAmount;
+            const takerAddress = await swapQuoteConsumerUtils.getTakerAddressOrThrowAsync(this.provider, opts);
+            const takerEthAndWethBalance = await swapQuoteConsumerUtils.getEthAndWethBalanceAsync(
+                this.provider,
+                this._contractWrappers,
+                takerAddress,
+            );
+            // TODO(david): when considering if there is enough Eth balance, should account for gas costs.
+            const isEnoughEthAndWethBalance = _.map(takerEthAndWethBalance, (balance: BigNumber) =>
+                balance.isGreaterThanOrEqualTo(ethAmount),
+            );
+            if (isEnoughEthAndWethBalance[1]) {
+                // should be more gas efficient to use exchange consumer, so if possible use it.
+                return this._exchangeConsumer;
+            } else if (isEnoughEthAndWethBalance[0] && !isEnoughEthAndWethBalance[1]) {
+                return this._forwarderConsumer;
+            }
+            // Note: Should we default to a consumer if takerAddress has not enough Eth or Weth?
```

Comment:
Maybe return the exchange consumer? Either way I agree we should return something and let that thing deal with the error. I would replace this comment with a comment that explains the reasoning.